### DISCUSSION TITLE
exclude_unset doesn't work if query-params are parsed as pydantic-model dependency

### DISCUSSION THREAD

**Discussion by solry** [fastapi/fastapi#9595] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from pydantic import BaseModel
from fastapi import FastAPI

app = FastAPI()

class TestModel(BaseModel):
    field_true: bool | None = Field(True)

@app.get("/test-unset")
def test_unset(
    model_depends_annotated: Annotated[TestModel, Depends(TestModel)],
    model_depends_non_annotated: TestModel = Depends(),
    model_non_depends: TestModel = TestModel(),
):
    return {
        "model_depends_annotated": model_depends_annotated.dict(
            exclude_unset=True,
        ),
        "model_depends_non_annotated": model_depends_non_annotated.dict(
            exclude_unset=True,
        ),
        "model_non_depends": model_non_depends.dict(
            exclude_unset=True,
        ),
    }
```

### Description

* Run your app, e.g. `python -m uvicorn --host 0.0.0.0 --port 8080 app:app`
* Open browser and call endpoint `http://localhost:8080/test-unset`
* Discover the following output:
```json
{
  "model_depends_annotated": {"field_true": true},
  "model_depends_non_annotated": {"field_true": true},
  "model_non_depends": {}
}
```
* When the expected output is
```json
{
  "model_depends_annotated": {},
  "model_depends_non_annotated": {},
  "model_non_depends": {}
}
```
* Conclude that the object that is shipped by the dependency injection doesn't have information which fields were set and not set.

### Operating System

macOS

### FastAPI Version

0.95.2

### Python Version

Python 3.10.8

### Additional Context

It could be related to https://github.com/tiangolo/fastapi/issues/5891, but the same result if I make `test_unset` an async function.

**Comment by starryknight64** [comment]
Verified this problem in Python 3.10.11, pydantic 1.10.18, and fastapi 0.95.2 (and uvicorn 0.31.0 in case it matters).

Also verified this problem still exists when updating to Python 3.12.7, pydantic 2.9.2, and fastapi 0.115.0 (and uvicorn 0.31.0 still).

I also tried calling `model_dump(exclude_unset=True)` directly in Python 3.12 (since the `dict` function is deprecated) and same results.

I also tried by setting `response_model_exclude_unset=True` in the decorator for the `test_unset` function (same results).

I could be wrong, but in my limited experience from what I can tell this looks like a Pydantic issue?

**Comment by YuriiMotov** [comment]
### My attempt to explain why it doesn't work:

<details>

When you use the syntax like
```python
model_depends_annotated: Annotated[TestModel, Depends(TestModel)]
```
or
```python
model_depends_non_annotated: TestModel = Depends()
```

it's actually the same as:

```python
model_depends_annotated: Annotated[TestModel, Depends(TestModel.__init__)]
```

where `TestModel.__init__` is automatically generated by Pydantic as something like:

```python
def TestModel.__init__(self, field_true: bool | None = Field(True)):
    self.field_true = field_true
```

FastAPI further analyzes the signature of `TestModel.__init__` and adds each parameter as separate dependency.

So, for FastAPI it's equal to the following code:

```python
async def get_test_model(field_true: bool | None = Field(True)):
    return TestModel(field_true=field_true)

@app.get("/test-unset")
def test_unset(model_depends_annotated: Annotated[TestModel, Depends(get_test_model)]):
    pass
```

As you can see, FastAPI will first validate the `field_true` parameter and then just pass it to the constructor of `TestModel`. So, this field will always be set (passed explicitly by FastAPI to model's constructor).

Since it's set explicitly, you can't exclude it with `exclude_unset=True`..

</details>

### Current situation

In FastAPI `0.115.0` the support of Pydantic models for `Query` (and also for `Header` and `Cookie`) parameters was introduced officially. So, the code should be changed to:

```python
from typing import Annotated
from pydantic import BaseModel, Field
from fastapi import FastAPI, Query

app = FastAPI()

class TestModel(BaseModel):
    field_true: bool | None = Field(True)

@app.get("/test-unset")
def test_unset(
    model_query: Annotated[TestModel, Query()],
):
    return {
        "model_query": model_query.model_dump(
            exclude_unset=True,
        ),
    }

```

But it still doesn't work as expected (unset values are not excluded):
```
curl -X 'GET' \
  'http://localhost:8000/test-unset' \
  -H 'accept: application/json'
{"model_query":{"field_true":true}}
```

**Reply by RichieB2B** [reply]
The above example works in FastAPI `0.127.1` when setting `None` as default:

```python
from typing import Annotated
from pydantic import BaseModel, Field
from fastapi import FastAPI, Query

app = FastAPI()

class TestModel(BaseModel):
    field_true: bool | None = None

@app.get("/test-unset")
def test_unset(
    model_query: Annotated[TestModel, Query()],
):
    return {
        "model_query": model_query.model_dump(
            exclude_unset=True,
        ),
    }
```
Now it works as expected:
```
curl -X 'GET' \
  'http://localhost:8000/test-unset' \
  -H 'accept: application/json'
{"model_query":{}}
```
The fact that it does not work when the default value is not None seems to be FastAPI specific (bug?). In plain pydantic this does work:
```python
from pydantic import BaseModel, Field

class TestModel(BaseModel):
    field_true: bool | None = Field(True)

test_model_not_set = TestModel()
print(f"test_model_not_set = {test_model_not_set.model_dump(exclude_unset=True)}")

test_model_set = TestModel(field_true=True)
print(f"test_model_set = {test_model_set.model_dump(exclude_unset=True)}")
```
This gives the output you'd expect:
```
test_model_not_set = {}
test_model_set = {'field_true': True}
```
